Solidity helper function for testing fixed points

VoR0220 opened the pull request:
One issue on this PR: As you will see in the upcoming PR, there is a test that will fail involving a passing in of a fixed point argument. Do we need to do something about getting the signature of a (u)fixed parameter? Shouldn't it be simply aliased to u256? (I believe that this is the problem)

Commit:
Solidity helper function for testing fixed points

Signed-off-by: VoR0220 <rj@erisindustries.com>

axic commented:
> Do we need to do something about getting the signature of a (u)fixed parameter?

I think that would be useful, to have`ufixed(M,N)` in the tests for clarity.

axic commented on test/libsolidity/SolidityExecutionFramework.h:
```diff
@@ -186,7 +187,12 @@ class ExecutionFramework
+	static u256 fixed(dev::bigint _numerator, dev::bigint _denominator, int _fixedBits)
+	{
+		rational _value = rational(dev::bigint(_numerator), dev::bigint(_denominator));
+		rational value = _value * boost::multiprecision::pow(bigint(2), _fixedBits);
```
Cannot you do this without having two variables?

VoR0220 replied:
This is how we create fixed point types in the current code...do you have a different suggestion?

axic replied:
I wonder if you can merge the multiplication into the first line?

VoR0220 replied:
ah, that's what you meant. I mean you could...but I think that would look very messy.

chriseth replied:
Wouldn't it be useful to also have this function overloaded for `double`? Not sure about rounding errors, though.

VoR0220 replied:
Yeah I don't think that's necessarily the best way to go, unless it was automatically given the `<< fixed` bit...somehow....probably doable.

axic replied:
> ah, that's what you meant. I mean you could...but I think that would look very messy.

Would this work?

``` diff
+       rational value = rational(dev::bigint(_numerator), dev::bigint(_denominator));
+       value *= boost::multiprecision::pow(bigint(2), _fixedBits);
```

axic replied:
@VoR0220 what you think re the above syntax? :)

axic reviewed (changes requested):
(Wanted to try out the new Github review system)

Please simplify to one intermediate variable.

axic commented:
@VoR0220 @chriseth can we finish and merge this?

VoR0220 commented:
@axic not yet. We need to get the inputs working properly. The outputs work. Just can't seem to figure out why the inputs won't take.

axic commented:
Inputs? Outputs? Are you sure we're talking about the same PR? :)

VoR0220 commented:
Yes, we are. So, when checking against the output, this fixed function works, but when trying to send arguments into the function call, this fails. I still don't know why this is.

chriseth commented:
@VoR0220 please explain what is not working in this pull request

VoR0220 commented:
@chriseth @axic So here is the problem in a nutshell. And it's a simple one. 

This does not work:

`BOOST_CHECK(callContractFunction("f(fixed,fixed)", u256(fixed(1, 3, 128)), u256(fixed(1, 2, 128))) == encodeArgs(fixed(1, 3, 128), fixed(1, 2, 128)));`

But this does:

`BOOST_CHECK(callContractFunction("f()") == encodeArgs(fixed(1, 32, 128)));`

I think that before merging this, the top should be confirmed working. I can't seem to get the top function to pass in the encoded fixed argument. Any and all suggestions are more than welcome.

axic commented:
@VoR0220 have you tried a simple one:

`BOOST_CHECK(callContractFunction("f(uint256)", u256(fixed(1, 32, 128)) == encodeArgs(u256(fixed(1, 32, 128))));`

where `function f(uint256 a) returns (uint) { return a; }`?

VoR0220 commented:
@axic no I haven't...but still...shouldn't it be set to the signature of fixed? Am I completely wrong here? I think as a hack this could work, but I really think this should be correct all out....I can try that approach though (and I bet it will work), but still...

axic commented:
@VoR0220 I think the point of this PR is not to implement `fixed`, but to add the helper method. I wouldn't be worried if those signatures are not working yet?

VoR0220 commented:
I suppose you're right. If @chriseth agrees, lets get this merged up.

chriseth committed:
Replace `fixed` by standard `encode` function.

This way, we can use
```
BOOST_CHECK(callContractFunction("t(fixed8x128)", make_pair(rational(2, 3), 128)) == encodeArgs(u256(9)));
```
which is a bit more self-explanatory, I would say.

chriseth commented:
@VoR0220 sorry for hijacking this pull request, but I really think we should keep this uniform.

axic commented:
I think the make_pair is a horrible syntax C++ has, but since there's nothing better and the overloading matches the rest, it should be better this way.
